ksql-common/src/main/java/io/confluent/ksql/processing/log/ProcessingLogConfig.java

```diff
@@ -24,6 +26,23 @@
 public class ProcessingLogConfig extends AbstractConfig {
   private static final String PROPERTY_PREFIX = "processing.log.";
 
+  private static ProcessingLogConfig INSTANCE = null;
```

Comment:
statics.... are nasty.  We should do everything we can to avoid having to rely on statics. Is there _really_ no alternative?

Reply:
Yeah, I don't love it either. I felt like in this case it might not be so bad since its usage is pretty contained, and its immutable (mostly, a test could still call configureFinal - couldn't think of a good way of preventing that), so you won't have the usual problems with unpredictable/untraceable behavior. The only things that need to modify the values are the processing log related tests, and we can pass a config in directly in those places. I was planning to enforce this more strictly (in a follow-up to keep the diff small) by moving everything that calls getInstance into another package (e.g. ksql.processing.log), and making getInstance package private.

The alternative would be to pass it all the way through to the  physical plan building stuff in the structured package. So we'd either have to:
    - Add these properties to KsqlConfig and pull them out of there, which I didn't want to do since KsqlConfig is becoming kind of a dumping ground for configs. We should refactor it to pull out the bits that track different AbstractConfig implementations into a parent manager class that can be configured to handle each implementation independently and be told which configs are compatibility-breaking or not.
     - Add it as a parameter everywhere.
     - Pass around a ConfigContext with the ksql config and this config.

I'm not super keen on any of the above, so I'm all ears to suggestions.

Reply:
KsqlConfig is a dumping ground - you're right.  But I don't think we're making the situation any better by ignoring it and using statics to get around the problem.  That's just added technical debt in my book.  

Maybe this is the right time to take a good look at our use of KsqlConfig?  Many methods already take both a `KsqlConfig` param and a `overrriddenProperties` param, and now we've got a processing log config.  

- `KsqlConfig` looks to be pretty much constant, i.e. the application creates it once at startup.  So why do we pass it around with each statement execution?  Could this not be injected at KsqlEngine creation?
- `overriddenProperties` can then be the bit that gets passed around per-execution.

That leaves the processing log properties.  Personally, I think we _should_ have a single top-level config class, `KsqlConfig` is probably a good name, but this should internally hold different instance of different config classes for the different areas, e.g. Ksql engine, streams and log processing.   

This top level class should also clearly define what the `overriddenProperties` should and should not be applied to. (i.e. SSL config and log processing properties should not be overridable).

In short, I feel that the log processing properties _should_ be part of some top level config.  Maybe now is the time to do a PR to refactor our current config usage to better support the addition of the processing log config...

What do you think? I'd even be willing to help if it meant we avoid statics! ;)

Reply:
> * `KsqlConfig` looks to be pretty much constant, i.e. the application creates it once at startup.  So why do we pass it around with each statement execution?  Could this not be injected at KsqlEngine creation?

Sure, but we still need to override with the persisted configs (for compatibility), then the user's  overrides before using it. So we have to pass around _something_, which IMO should be KsqlConfig. We can add a method for getting the overrides from it so we can return that info from that rest api (which is all they're used for after the call to `cloneWithPropertyOverwrites`. That way we don't have to pass the overrides around separately.
 
> Personally, I think we _should_ have a single top-level config class, `KsqlConfig` is probably a good name, but this should internally hold different instance of different config classes for the different areas

Agreed. This is pretty much that parent manager class I mentioned.

I don't want to block this change on the whole config refactor, since it should get into the next release. As a first step, I'll wrap the ksql config, overridden properties, and processing log config into another class, and we can pass that around. Then we can evolve that class into the thing that manages all the subsystem configs.

Reply:
I opted to pass around a context object for the processing log. Currently you can use the context to get to the logger factory and the config. In a follow-up I'll move the msg factories into it, and drop the public interface to get the config. So the pattern can just be that you initialize the context with the config when the process starts up, and use it to do all your processing log stuff. This also gets rid of the global logger factory.